### PR TITLE
fix leaking on conversion errors and panics

### LINKED ISSUE #549: implicit function param conversions leaks memory
extendr functions/method allow conveniently to set types in function param, to skip explicitly converting any Robj into something else. E.g. below `x: Strings` implicit call something like `robj.as_strings()` under the hood. However any conversion error will activate `throw_r_error`

`throw_r_error()` is easy to use, but known for leaking mem #278 . The history of alternatives are long and there is a broad spectrum of different opinions/patterns in the community. 

extendr relies on `throw_r_error` in this function header conversion. E.g. in this function Robj is converted to Strings. "use_try_from" can be true or false
```rust
#[extendr(use_try_from = true)]
fn give_me_string_vec(x: Strings) -> String {
    format!("thank you for Strings with length {} ...", x.len())
}
```

In this case the leakage is benign, only 3 integers or some Altrep thereof. For use cases where input is small, the user would find status quo completely acceptable.
```R
> polars:::give_me_string_vec(1:3)
Error in polars:::give_me_string_vec(1:3) : Expected Strings got Integers
```

however this will cause a 2Gb leak, which will not be recovered until Rsession is restarted
```R
(function() {
  for (i in 1:100) {
    print(i)
    x = sample(1:10,1e7,replace = TRUE)
    tryCatch(polars:::give_me_string_vec(x), error = \(err) NULL)
  }
})()
```

work arounds is to take any input as `Robj` and use `try_into` and return a result (#473) or panic!. Then there is no leakage.

### PR DESCRIPTION
**Pull request by sorhawell**
fixes #549

### CODE REVIEW

**Comment by sorhawell** [comment]
[test results](https://github.com/sorhawell/does_it_leak/blob/simple_example/tests/leak_result.txt).

The tiny 50-250 bytes leaks in first test is likely a burn-in fluke.

**Comment by sorhawell** [comment]
example of the new macro expansion:
image

File: extendr-macros/src/wrappers.rs [hunk 1]
```diff
@@ -102,17 +100,44 @@ pub fn make_function_wrappers(
...
     // ```
+    //
     wrappers.push(parse_quote!(
         #[no_mangle]
         #[allow(non_snake_case, clippy::not_unsafe_ptr_arg_deref)]
         pub extern "C" fn #wrap_name(#formal_args) -> extendr_api::SEXP {
-            unsafe {
-                use extendr_api::robj::*;
+            use extendr_api::robj::*;
+            let res_res: std::result::Result<
```
Comment (multimeric) [inline extendr-macros/src/wrappers.rs:109]: More meaningful name

File: extendr-macros/src/wrappers.rs [hunk 2]
```diff
@@ -102,17 +100,44 @@ pub fn make_function_wrappers(
...
+                std::panic::catch_unwind(||-> std::result::Result<Robj, extendr_api::Error> {
+                    Ok(extendr_api::Robj::from(#call_name(#actual_args)))
+                })
+            };
+            match res_res {
+                Ok(Ok(zz)) => {
+                    return unsafe { zz.get() };
+                }
+                Ok(Err(extendr_err)) => {
+                    let err_string = extendr_err.to_string();
```
Comment (multimeric) [inline extendr-macros/src/wrappers.rs:123]: I'm confused as to how the `Ok(Err)` path occurs. Won't it always be `Err()` or `Ok(Ok())` given this code?

**Comment by sorhawell** [comment]
> Hmm, this confuses me because it really seems like the only thing that this changes is adding `drop(extendr_err)` which is basically what #551 did. Am I missing something?

The difference is unwind and throw_r_error is very dodgy way to do clean up memory and even though all objects is dropped in one scope, there could be other objects in the nested tree of scopes which aren't dropped when doing longjump.

A quite safer/cleaner way is to return a nested result object from the unsafe scope. Then guaranteed by the rust scope rules, anything but the result-object is then dropped. Then only the result object in a single scope needs to be dealt with  This makes it easier to get it right.

See also [comment in 551](https://github.com/extendr/extendr/pull/551#issuecomment-1556854052).

.. I should though provide a link to the tests 551 fails

File: extendr-macros/src/wrappers.rs (hunk 2 above)
Comment (sorhawell) [reply extendr-macros/src/wrappers.rs:123]: `Err(Any)` contains any caught panic. `Ok(Err(Error))` contains any conversion error. Since on new line 330 I replace `unwrap_or_throw_error` with ?-operator to bubble error instead of throwing and panicking. There for the expansion of `#actual_args` has one `?` for each parameter. When a conversion fails  will return from closure on line 114.  Notice the type on line 114 is return type of the closure, not from `catch_unwind()`.

Comment (sorhawell) [reply extendr-macros/src/wrappers.rs:123]: if some user function itself returns a Result-type which is Err and since something like #473 is not adopted, `extendr_api::Robj::from` can then also panic in the post-conversion back to R. I define this also as a user induced panic. So a user function will only panic if the user uses panic/unwrap/... or if returning an Err.

Comment (sorhawell) [reply extendr-macros/src/wrappers.rs]: image

Comment (sorhawell) [reply extendr-macros/src/wrappers.rs]: looking at the expansion again, I see can drop the `map_err(|e| ...)` as `?`-operator does that implicitly.
image

Comment (yutannihilation) [reply extendr-macros/src/wrappers.rs]: > I'm confused as to how the `Ok(Err)` path occurs. 

Yeah, this is tricky, but this has to be like this because we have to handle `panic!()`.

Just for information, Lionel Henry kindly suggested that I could write some C code to wrap `R_UnwindProtect()` so that no `panic!()` is needed to escape from the cleanup function. I'll experiment this idea and share the result (it probably won't be very soon, so please forget for now).

**Comment by sorhawell** [comment]
@Ilia-Kosenkov @multimeric 

I believe this PR is ready to merge and all conversations are resolved.

I think it is central to extendr to solve the memory leaking now.

**Comment by CGMossa** [comment]
I'll review it soon.

**Comment by CGMossa** [comment]
Does this **actually** fix the leaks?

I've no comments to add.. But I cannot approve this because there are no tests,
and no message from you that state that this branch, doesn't leak memory.
You even made the leak tests for this! Am I missing something?

(sorry I had food poisoning)

**Comment by multimeric** [comment]
Agreed that tests are important here. Could the R benchmarks be converted into unit tests?

**Comment by CGMossa** [comment]
@multimeric okay, I might understand this now: This message from @sorhawell state that he ran tests on this PR, master, and #551. https://github.com/extendr/extendr/pull/555#issuecomment-1557134693 And this PR passes everything.

@sorhawell I think you should just put these tests in this PR as well.

**Comment by multimeric** [comment]
Yeah, I know he tested this externally, it's just that including them in the PR increases our confidence about it, and prevents regressions.

**Comment by Ilia-Kosenkov** [comment]
I am having trouble mapping the expanded macro from the screenshot above to the actual code right now. Would you be able to show us an expanded macro of a function with, say, three different args?

**Comment by sorhawell** [comment]
> I am having trouble mapping the expanded macro from the screenshot above to the actual code right now. Would you be able to show us an expanded macro of a function with, say, three different args?

One way to see the expansion is to e.g. use helloextendr or similar. Refer to my repo + branch in Cargo.toml. Then run `cargo expand` .

I will try to post an example at some point.

**Comment by sorhawell** [comment]
@Ilia-Kosenkov  I have made an example from helloextendr with 3 args and use_try_from  = true and false

https://github.com/sorhawell/three_arg_expansion/blob/main/README.md

File: extendr-macros/src/wrappers.rs [hunk 3]
```diff
@@ -102,17 +100,48 @@ pub fn make_function_wrappers(
...
+
+
+            match wrap_result_state {
+                Ok(Ok(zz)) => {
+                    return unsafe { zz.get() };
+                }
+                // any conversion error bubbled from #actual_args conversions of incomming args from R.
+                Ok(Err(conversion_err)) => {
+                    let err_string = conversion_err.to_string();
+                    drop(conversion_err); // try_from=true errors contain Robj, this must be dropped to not leak
```
Comment (Ilia-Kosenkov) [inline extendr-macros/src/wrappers.rs:128]: I may be missing the context, but why is it not dropped at the end of the scope (as all Rust things are)? Is it because `::throw_r_error()` breaks things up?

Comment (sorhawell) [reply extendr-macros/src/wrappers.rs:128]: Paraphrasing the writings of @yutannihilation . Rust has a stack and perhaps some heap allocations. Performing longjump will not unwind the stack and free heap allocations. But also panic unwind is discouraged in rust docs and stated as UB when returning to another language as R.

Previous attempts to manually drop before throw_r_error() did not fully succeed because they only dropped one or some scopes on top of the stack. By first returning `Result` the whole conversion stack can be dropped, and then only a single object needs to be dealt with. That is a saner way to deal with all edge cases and it emperically solves any known case of leaking out of ~150 test cases.

I have used with r-polars and there is none of 1200 unit tests which fail.

**Comment by Ilia-Kosenkov** [comment]
So what is the consensus here? Do we need integration tests or perhaps separate benchmarks in the project?

**Comment by sorhawell** [comment]
I was looking extendr tests yesterday, and was thinking to place one or two extendr-impls like  [here](https://github.com/extendr/extendr/blob/master/tests/extendrtests/src/rust/src/optional_either.rs)

and then write a test in testthat folder.

**Comment by sorhawell** [comment]
I have added the benchmark as integration test and reduced testing from ~150 to 20 tests.

**Comment by yutannihilation** [comment]
Yes, as I wrote "except for Windows," nothing new.

**Comment by sorhawell** [comment]
I do not plan to add more to this PR for now.
For my own pet project I will adopt 555, 553 and 473.

**Comment by Ilia-Kosenkov** [comment]
Sorry, I was overwhelmed at work recently and did not check these PRs. I will review them & we will hopefully merge.

**Comment by sorhawell** [comment]
@yutannihilation 
Now when this [other unrelated win-build error](https://github.com/extendr/extendr/actions/runs/5200353819/jobs/9378963508) ever went away, what is your opinion on merging this PR?

**Comment by yutannihilation** [comment]
I'm in [a different direction](https://github.com/extendr/extendr/pull/555#discussion_r1207849111), so I'm not the right one to judge this, but, to me, there seems no reason not to merge this pull request. It looks far better than the current code.

**Comment by CGMossa** [comment]
@multimeric should we just merge this then?

**Review by Ilia-Kosenkov — APPROVED** [review]
This should also be mentioned in `CHANGELOG.md`

**Comment by eitsupi** [comment]
@sorhawell Could you update the changelog?

**Comment by sorhawell** [comment]
@Ilia-Kosenkov  changelog updated

**Comment by CGMossa** [comment]
Thanks @Ilia-Kosenkov

**Comment by CGMossa** [comment]
@yutannihilation Interesting, have you seen this https://www.tidyverse.org/blog/2019/05/resource-cleanup-in-c-and-the-r-api/ ? Did you already link to it?

**Comment by yutannihilation** [comment]
No.
